zerver/lib/storage.py

```diff
@@ -24,26 +24,63 @@ def static_path(path: str) -> str:
         return os.path.join(settings.STATIC_ROOT, path)
 
 
-def reformat_medium_filename(hashed_name: str) -> str:
-    """
-    Because the protocol for getting medium-size avatar URLs
-    was never fully documented, the mobile apps use a
-    substitution of the form s/.png/-medium.png/ to get the
-    medium-size avatar URLs. Thus, we must ensure the hashed
-    filenames for system bot avatars follow this naming convention.
-    """
-
-    name_parts = hashed_name.rsplit(".", 1)
-    base_name = name_parts[0]
-
-    if len(name_parts) != 2 or "-medium" not in base_name:
-        return hashed_name
-    extension = name_parts[1].replace("png", "medium.png")
-    base_name = base_name.replace("-medium", "")
-    return f"{base_name}-{extension}"
+class IgnoreBundlesManifestStaticFilesStorage(ManifestStaticFilesStorage):
+    hashed_static_avatar_file_map: dict[str, str] = {}
 
+    def process_static_avatars_name(
+        self,
+        name: str,
+        content: Optional["File[bytes]"] = None,
+        filename: str | None = None,
+    ) -> str:
+        """
+        Because the protocol for getting medium-size avatar URLs
+        was never fully documented, the mobile apps use a
+        substitution of the form s/.png/-medium.png/ to get the
+        medium-size avatar URLs.
+
+        This function hashes system bots' avatar files in a way
+        that follows the pattern used for user-uploaded avatars.
+
+        It ensures the following:
+
+            * Hashed filenames for system bot avatars follow this
+            naming convention:
+            - avatar.png -> avatar-medium.png
+
+            * The system bots' default avatar file and its medium
+            version share the same hash:
+            - bot.36f721bad3d0.png -> bot.36f721bad3d0-medium.png
+        """
+
+        def reformat_medium_filename(hashed_name: str) -> str:
+            name_parts = hashed_name.rsplit(".", 1)
+            base_name = name_parts[0]
+
+            if len(name_parts) != 2 or "-medium" not in base_name:
+                return hashed_name
+            extension = name_parts[1].replace("png", "medium.png")
+            base_name = base_name.replace("-medium", "")
+            return f"{base_name}-{extension}"
+
+        if name.endswith("-medium.png"):
+            # This logic relies on the fact that the medium files will
+            # be hashed first due to the "-medium" string, which places
+            # them earlier in the naming order. So, adhering to the
+            # medium file naming convention is crucial for this logic.
```

Comment:
I wonder if we need a more robust logic than the one described here, one that can produce the same result no matter the file order. But considering this is only used by system bots and doesn't break things much if wrong, I opted to implement  simpler logic

Reply:
I think the correct thing to do here is to write an `assert` statement that will fail if the `-medium` file hasn't already been processed as thus added an appropriate entry to `hashed_static_avatar_file_map`.